src/storage/v2/storage_error.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include "storage/v2/constraints.hpp"
+
+#include <variant>
+
+namespace memgraph::storage {
+
+enum class ReplicationError {
+  UNABLE_TO_SYNC_REPLICATE,
+};
+
+struct StorageDataManipulationError {
+  std::variant<ConstraintViolation, ReplicationError> error;
+
+  bool operator==(const StorageDataManipulationError &) const = default;
+};
+
+enum class IndexDefinitionError {
+  EXISTANT_INDEX,
+  NONEXISTANT_INDEX,
+};
+
+struct StorageIndexDefinitionError {
+  std::variant<IndexDefinitionError, ReplicationError> error;
+
+  bool operator==(const StorageIndexDefinitionError &) const = default;
+};
+
+enum class ConstraintDefinitionError {
+  EXISTANT_CONSTRAINT,
+  NONEXISTANT_CONSTRAINT,
+};
+
+struct StorageExistenceConstraintDefinitionError {
+  std::variant<ConstraintViolation, ConstraintDefinitionError, ReplicationError> error;
+
+  bool operator==(const StorageExistenceConstraintDefinitionError &) const = default;
+};
+
+struct StorageExistenceConstraintDroppingError {
+  std::variant<ConstraintDefinitionError, ReplicationError> error;
+
+  bool operator==(const StorageExistenceConstraintDroppingError &) const = default;
+};
+
+struct StorageUniqueConstraintDefinitionError {
+  std::variant<ConstraintViolation, ReplicationError> error;
+
+  bool operator==(const StorageUniqueConstraintDefinitionError &) const = default;
+};
+
+struct StorageUniqueConstraintDroppingError {
+  std::variant<ReplicationError> error;
+
+  bool operator==(const StorageUniqueConstraintDroppingError &) const = default;
+};
```

Comment:
Sure. 
The errors look all similar: they are all variant of something and `ReplicationError`. However the difference comes from the extra 'something': for instance when creating (or dropping) an `ExistenceConstraint` you might get the following variant:
-`ReplicationError` (sync replica couldn't confirm)
-`ConstraintDefinitionError`: is the constraint already existing or not. This was existing before this PR but distinction was directly done in the interpreter.
-`ConstraintViolation`: is the constraint already violated by existing nodes. Same remark than `ConstraintDefinitionError`.

There might be a way to have something even more generic but it will make it less clear to my opinion. It took me a while to implement the whole variant/error aspect as I needed to really dive in the code, make sure which workflow was followed for each query, not break the existing behavior etc... I hope this approach would make it easier if anyone needs to come and correct something in the future.

Reply:
I mean why should we have a separate error for unique constraints and existence constraints? So instead of having both `StorageExistenceConstraintDroppingError` and `StorageUniqueConstraintDroppingError`, I would only have `StorageConstraintDroppingError`, because in my mind it makes sense to get the same error similarly to `ConstraintViolatioin`.

Reply:
After discussing this with @42jeremy I realized they are not the same. The existence constraint related errors can hold a `ConstraintDefinitionError`, while the unique constraints cannot.